server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,443 @@
+        """
+        self.description = _('Rsync files to remote destination')
+        super(PublishStep, self).__init__(step_type, config=config)
+        self.file_list = file_list
+        self.fast_forward = fast_forward
+        self.exclude = exclude
+        self.delete = delete
+        self.src_directory = src_directory
+        self.dest_directory = dest_directory
+        self.links = links
+
+    def remote_mkdir(self, path):
+        """
+        Creates path on remote server. The path is rooted in distributor's remote_root directory.
+
+        :param path: path to create on remote server
+        :type path: str
+        :return: (return code from rsync, stdout + stderr)
+        :rtype: tuple
+        """
+        tmpdir = os.path.join(self.get_working_dir(), '.tmp')
+        os.makedirs(os.path.join(tmpdir, path.lstrip("/")))
+        args = ['rsync', '-avr', '-f+ */']
+        args.extend(self.make_authentication())
+        args.append("%s/" % tmpdir)
+        args.append(self.make_destination(path).replace(str(path), ""))
+        _logger.debug("remote mkdir: %s" % args)
+        is_ok, output = self.call(args, include_args_in_output=True)
+        _logger.debug("remote mkdir out: %s" % output)
+        shutil.rmtree(tmpdir)
+        return is_ok, output
+
+    def make_ssh_cmd(self, args=None):
+        """
+        Returns a list of arguments needed to form an ssh command for connecting to remote server.
+
+        :param args:list of extra args to append to the standard ssh command
+        :type args: list
+        :return: list of arguments for ssh portion of command
+        :rtype: list
+        """
+        user = self.get_config().flatten()["remote"]['ssh_user']
+        auth_type = self.get_config().flatten()["remote"]['auth_type']
+
+        # -e 'ssh -l ssh_user -i ssh_identity_file'
+        # use shared ssh connection for other threads
+        cmd = ['ssh', '-l', user]
+        if auth_type == "publickey":
+            key = self.get_config().flatten()["remote"]['ssh_identity_file']
+            cmd += ['-i', key]
+        cmd += ['-o', 'StrictHostKeyChecking no',
+                '-o', 'UserKnownHostsFile /dev/null',
+                '-S', '/tmp/rsync_distributor-%r@%h:%p',
+                '-o', 'ControlMaster auto',
+                '-o', 'ControlPersist 10']
+        if self.get_config().flatten()["remote"]['auth_type'] == 'password':
+            password_file = os.path.join(self.get_working_dir(), str(uuid.uuid4()))
+            open(password_file, 'w').write(self.get_config().flatten()['remote']['ssh_password'])
+            cmd = ['sshpass', '-f', password_file] + cmd
+        if args:
+            cmd += args
+        return cmd
+
+    def make_authentication(self):
+        """
+        Returns a list of strings representing args for command for authenticating against a remote
+        server.
+
+        :return: list of arguments for auth. e.g., ['-e',  ssh -l ssh_user -i /key_path'
```

Comment:
This list example is invalid. The quoting isn't complete and there there isn't a closing list bracket at the end.